doc/designs/ech-api.md

```diff
@@ -0,0 +1,711 @@
+is required.
+
+```c
+typedef unsigned int (*SSL_ech_cb_func)(SSL *s, const char *str);
+
+void SSL_ech_set_callback(SSL *s, SSL_ech_cb_func f);
+void SSL_CTX_ech_set_callback(SSL_CTX *ctx, SSL_ech_cb_func f);
+```
+
+The following options are defined for ECH and may be set via
+`SSL_set_options()`:
+
+```c
+/* set this to tell client to emit greased ECH values when not doing
+ * "real" ECH */
+#define SSL_OP_ECH_GREASE                               SSL_OP_BIT(36)
+/* If this is set then the server side will attempt trial decryption */
+/* of ECHs even if there is no matching record_digest. That's a bit  */
+/* inefficient, but more privacy friendly */
+#define SSL_OP_ECH_TRIALDECRYPT                         SSL_OP_BIT(37)
+/* If set, clients will ignore the supplied ECH config_id and replace
+ * that with a random value */
+#define SSL_OP_ECH_IGNORE_CID                           SSL_OP_BIT(38)
+/* If set, servers will add GREASEy ECHConfig values to those sent
+ * in retry_configs */
+#define SSL_OP_ECH_GREASE_RETRY_CONFIG                  SSL_OP_BIT(39)
+```
+
+A Note on `_get_`,`_get0_`,`_get1_`,`_set_`,`_set0_`,`_set1_`
+-------------------------------------------------------------
+
+TODO(ECH): This text will likely disappear as things settle.
+
+The abstraction behind the `_get_`,`_get0_`,`_get1_`,`_set_`,`_set0_`,`_set1_`
+convention used in OpenSSL APIs is somewhat non-obvious, (but is what it is),
+so some words of explanation of the function names above may be useful, partly
+as a check that those usages are consistent with other APIs:
+
+- `_set_` is appropriate where the input/output type(s) are basic and involve
+  no type-specific memory management (e.g. `SSL_set_enable_ech_grease`)
+- there are no uses of `_get_` or `_get0_` above
+- `_get1_` is appropriate when a pointer to a complex type is being returned
+  that may be modified and must be free'd by the application, e.g.
+  `OSSL_ECHSTORE_get1_info`.
+- `_set0_` is also unused above, because...
+- the `_set1_` variant seems easier to handle for the application ("with ECH
+  stuff, if you make it then give it to the library, you still need to free
+  it") and for consistency amongst these APIs, so that is often used, e.g.
+  `OSSL_ECHSTORE_set1_key_and_list`.
+
+Build Options
+-------------
+
+All ECH code is protected via `#ifndef OPENSSL_NO_ECH` and there is
+a `no-ech` option to build without this code.
+
+BoringSSL APIs
+--------------
+
+Brief descriptions of boringssl APIs are below together with initial comments
+comparing those to the above. (It may be useful to consider the extent to
+which it is useful to make OpenSSL and boring APIs resemble one another.)
+
+Just as our implementation is under development, boring's `include/openssl/ssl.h`
+says: "ECH support in BoringSSL is still experimental and under development."
+
+### GREASE
+
+Boring uses an API to enable GREASEing rather than an option.
```

Comment:
Not sure we need to draw a comparison to boringssl here, but if we do, we should probably use the fully qualified name.